Frontend/nix/watch.nix

```diff
@@ -0,0 +1,115 @@
+    ../node
+  ];
+
+  perSystem = { config, self', lib, pkgs, system, ... }:
+    let
+      webpack-dev-server = pkgs.writeShellApplication {
+        name = "webpack-dev-server-wrapped";
+        runtimeInputs = [ self'.packages.nodejs ];
+        text = ''
+          export NODE_PATH="${self'.packages.nodeDependencies}/node_modules"
+          PIPE="$1"
+          shift
+          tail -n1 -f "$PIPE" | node ${./watch.js} "$@"
+        '';
+      };
+
+      purifix-watch =
+        pkgs.writeShellApplication {
+          name = "purifix-watch";
+          runtimeInputs = [
+            pkgs.watchexec
+          ] ++ config.packages.ui-common.develop.buildInputs;
+          text = ''
+            PIPE=$1
+            shift
+            SRC_FOLDERS=()
+            SRC_GLOBS=()
+            for SRC_FOLDER in "$@"; do
+              if [ -d "$SRC_FOLDER/src" ]; then
+                SRC_FOLDERS+=( -w "$SRC_FOLDER/src" )
+                SRC_GLOBS+=( "\"$SRC_FOLDER/src/**/*.purs\"" )
+              fi
+            done
+            watchexec "''${SRC_FOLDERS[@]}" -r -f "**/*.purs" -- \
+              "echo 'suspend' >$PIPE;purifix ''${SRC_GLOBS[*]} && echo 'ready' >$PIPE"
+          '';
+        };
+
+      make-watch-processes = { target, env, platform }:
+        let
+          pipe = "/tmp/watch-${target}-${platform}-${env}";
+        in
+        {
+          port = 7812;
+          tui = false;
+          settings.processes = {
+            make-pipe = {
+              command = "mkfifo ${pipe}";
+              is_daemon = true;
+              shutdown.command = "echo 'Shutting down'; rm -f ${pipe}";
+              availability.restart = "exit_on_failure";
+              readiness_probe = {
+                exec.command = "[ -e ${pipe} ]";
+                period_seconds = 1;
+              };
+            };
+
+            purifix-watch = {
+              command = ''
+                cd ./Frontend/${target}
+                ${purifix-watch}/bin/purifix-watch ${pipe} . ../ui-common
+              '';
+              depends_on.make-pipe.condition = "process_healthy";
+            };
+
+            webpack-watch = {
+              command = ''
+                ${webpack-dev-server}/bin/webpack-dev-server-wrapped \
+                    ${pipe} \
+                    --env ${env}   \
+                    --config "$(realpath ./Frontend/${target}/webpack.${platform}.js)" \
+                    --entry ./Frontend/${target}/index.js
+              '';
+              depends_on.make-pipe.condition = "process_healthy";
+            };
+          };
+        };
+
+      process-options =
+        lib.attrsets.cartesianProductOfSets {
+          env = [ "master" "sandbox" "prod" ];
+          target = [ "ui-customer" "ui-driver" ];
+          platform = [ "android" ]; # TODO: support iOS
+        };
+
+      process-name = { target, platform, env }: "${target}-start-${platform}-${env}";
+
+      process-configs =
+        builtins.listToAttrs (map
```

Comment:
There are quite a few places -- here and in `bundle.nix` -- you are using `builtins.listToAttrs . map`. This is unnecessary when your build configurations info is essentially an attrset. What you want is something like:

```nix
let
   build-configs =
      let choices = lib.attrsets.cartesianProductOfSets {
          env = [ "master" "sandbox" "prod" ];
          target = [ "ui-customer" "ui-driver" ];
          platform = [ "android" ]; # TODO: support iOS
        };
     in builtins.listToAttrs (map (args: { name = "${args.target}-start-${args.platform}-${args.env}"; value = args; }) choices);
```

Then you can just `lib.mapAttrs _ build-configs` which is far more straightforward to understand on first glance.